pkg/scale/decode.go

```diff
@@ -478,59 +478,87 @@ func (ds *decodeState) decodeBool(dstv reflect.Value) (err error) {
 
 // decodeUint will decode unsigned integer
 func (ds *decodeState) decodeUint(dstv reflect.Value) (err error) {
-	b, err := ds.ReadByte()
+	const maxUint32 = ^uint32(0)
+	const maxUint64 = ^uint64(0)
+	prefix, err := ds.ReadByte()
 	if err != nil {
-		return
+		return fmt.Errorf("reading from buffer: %w", err)
 	}
 
 	in := dstv.Interface()
 	temp := reflect.New(reflect.TypeOf(in))
 	// check mode of encoding, stored at 2 least significant bits
-	mode := b & 3
-	switch {
-	case mode <= 2:
-		var val int64
-		val, err = ds.decodeSmallInt(b, mode)
+	mode := prefix % 4
+	var value uint64
+	switch mode {
+	case 0:
+		value = uint64(prefix >> 2)
+	case 1:
+		buf, err := ds.ReadByte()
 		if err != nil {
-			return
+			return fmt.Errorf("reading from buffer: %w", err)
 		}
-		temp.Elem().Set(reflect.ValueOf(val).Convert(reflect.TypeOf(in)))
-		dstv.Set(temp.Elem())
-	default:
-		// >4 byte mode
-		topSixBits := b >> 2
-		byteLen := uint(topSixBits) + 4
-
+		value = uint64(binary.LittleEndian.Uint16([]byte{prefix, buf}) >> 2)
+		if value <= 0b0011_1111 || value > 0b0111_1111_1111_1111 {
+			return fmt.Errorf("%w: %d (%b)", ErrU16OutOfRange, value, value)
+		}
+	case 2:
+		buf := make([]byte, 3)
+		_, err = ds.Read(buf)
+		if err != nil {
+			return fmt.Errorf("reading from buffer: %w", err)
```

Comment:
nit for `Decoder`s, it's just an `io.Reader` so not necessarily a buffer. Also we're reading from the `io.Reader` to the 3-bytes-buffer

```suggestion
			return fmt.Errorf("reading bytes: %w", err)
```